### ISSUE TITLE
Backend response of 426 Upgrade Required results in a 502

### ISSUE THREAD

**Issue by ronaldwind** [haproxy/haproxy#1997]
### Detailed Description of the Problem

In case our backend returns a 426 response, HAProxy errors and returns a 502.

This is the error HAProxy produces:

```
$ echo "show errors" | socat stdio UNIX-CONNECT:/var/run/haproxy/admin.sock

[10/Jan/2023:14:15:24.144] backend backend_api (#10): invalid response
  frontend frontend_https (#4), server 20230110-104426-Node4 (#4010007), event #3080, src x.x.x.x:13316
  buffer starts at 0 (including 0 out), 16186 free,
  len 198, wraps at 16336, error at position 198
  H1 connection flags 0x80002200, H1 stream flags 0x00004810
  H1 msg state MSG_DONE(30), H1 msg flags 0x00009595
  H1 chunk len 0 bytes, H1 body len 0 bytes :

  00000  HTTP/1.1 426 Upgrade Required\r\n
  00031  cache-control: max-age=0, private, must-revalidate\r\n
  00083  connection: upgrade\r\n
  00104  content-length: 0\r\n
  00123  date: Tue, 10 Jan 2023 14:15:23 GMT\r\n
  00160  server: Cowboy\r\n
  00176  upgrade: websocket\r\n
  00196  \r\n
```

This results in a 502 being sent as the response to the connecting client.

### Expected Behavior

I would expect the backends response of the 426 to be passed to the connecting http client. And I don't expect haproxy to encounter an error as retrieved by `show errors`.

### Steps to Reproduce the Behavior

In case the expected behavior is indeed what I could expect from HAProxy, then I'll try to create a reproduction plan. In the end it's a matter of having a backend send out the 426 with the headers as found in the error mentioned above.

### What is your configuration?

```haproxy
global
        log 127.0.0.1:9514 local0
        log /dev/log    local0
        log /dev/log    local1 notice
        chroot /var/lib/haproxy
        stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
        stats timeout 30s
        user haproxy
        group haproxy
        daemon
        maxconn 8000
        hard-stop-after 5s

        # Default SSL material locations
        ca-base /etc/ssl/certs
        crt-base /etc/ssl/private

        ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
        ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
        ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets
        ssl-default-server-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
        ssl-default-server-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
        ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets
        tune.ssl.default-dh-param 4096
        ssl-dh-param-file /etc/haproxy/dhparams.pem

defaults
        log     global
        mode    http
        option  dontlognull
        timeout connect 6000
        timeout client  60000
        timeout server  60000
        compression algo gzip
        compression type text/plain text/css application/x-javascript text/xml application/xml application/xml+rss text/javascript
        errorfile 400 /etc/haproxy/errors/400.http
        errorfile 403 /etc/haproxy/errors/403.http
        errorfile 408 /etc/haproxy/errors/408.http
        errorfile 500 /etc/haproxy/errors/500.http
        errorfile 502 /etc/haproxy/errors/502.http
        errorfile 503 /etc/haproxy/errors/503.http
        errorfile 504 /etc/haproxy/errors/504.http
        log-format "%ci:%cp [%tr] %ft %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs %{+Q}r [pid:%pid]"

peers mypeers
        peer Node1 Node1:9999
        peer Node2 Node2:9999
        peer Node4 Node4:9999

frontend frontend_https
        bind *:443 ssl crt /opt/azure-vault/secrets/star-example-com.pem ecdhe secp384r1 alpn h2,http/1.1
        tcp-request inspect-delay 5s
        option http-buffer-request
        use_backend backend_api

backend backend_api
        balance roundrobin
        server 20230110-104426-Node2 Node2:10007 id 2010007 check weight 100
        server 20230110-104426-Node4 Node4:10007 id 4010007 check weight 100
        option forwardfor
        http-request set-header X-Real-IP %[src]
        http-request set-header X-Forwarded-Host %[req.hdr(Host)]
        http-request set-header X-Forwarded-Port %[dst_port]
        http-request add-header X-Forwarded-Proto https if { ssl_fc }
        http-response set-header Server haproxy
        http-response set-header Strict-Transport-Security max-age=63072000
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.20-1ppa1~jammy 2022/12/09 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.20.html
Running on: Linux 5.15.0-1024-azure #30-Ubuntu SMP Wed Nov 16 23:37:59 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_OT=1 USE_PROMEX=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS +OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
Running on OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with OpenTracing support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.39 2021-10-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=

Available services : prometheus-exporter
Available filters :
        [  OT] opentracing
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

**Comment by capflam** [comment]
Indeed, there is a bug. `Upgrade` header on the response is handled in the same way, independently on the status code. It means that for a `426-Upgrade-required` response, haproxy will trigger a parsing error because there is no websocket key header.

I will push a fix soon. Thanks !

**Comment by T4cC0re** [comment]
Coincidentally I came across the same issue today as well on HAProxy 2.6 upgrading from HAProxy 1.x.

Is it planned to cut the fix for this issue into a release within a reasonably short timeframe?

**Comment by wtarreau** [comment]
Indeed, Upgrade should only be parsed on 101 because this is the only valid status code to perform a protocol upgrade.

**Comment by capflam** [comment]
It should be fixed in 2.8-DEV

**Comment by T4cC0re** [comment]
Is there a timeline on when this fix will be backported and released to the current 2.6 LTS?

**Comment by capflam** [comment]
Well, we planned to emit a bunch of releases this week. So I would say it will be very soon :)
